app/controllers/spree/orders_controller.rb

```diff
@@ -215,5 +216,17 @@ def order_params
         line_items_attributes: [:id, :quantity]
       )
     end
+
+    def process_payment_intent!(payment_intent)
+      return unless payment_intent&.starts_with?("pi_")
+      return unless order = Spree::Order.find_by!(number: params[:id])
+
+      last_payment = OrderPaymentFinder.new(order).last_payment
+      return unless last_payment&.state == "pending" &&
+                    last_payment&.response_code == payment_intent
+
+      last_payment.update_attribute(:cvv_response_message, nil)
+      last_payment.complete!
+    end
```

Comment:
IMO this should be extracted into a `ProcessPaymentIntent` service so we don't couple the controller with intents' low-level details, which also give us room to extract private helper methods that will make this code more readable. I imagine something like 

```rb
def initialize(order, payment_intent)
end

def call
  return unless valid?

  last_payment.update_attribute(:cvv_response_message, nil)
  last_payment.complete!
end

private

def valid?
  return false unless valid_intent_string?
  return false unless matches_last_payment?
  true
end
```

Reply:
891874995bd3cddf310b6e73ec851789227f4d33 👍